README.md
![alt text](./docs/_static/logo.png)

### The minimalist - all purpose - color class for printing
[PyPi](https://pypi.org/project/object-colors/) |
[Read the Docs](https://object-colors.readthedocs.io/en/latest/) |
[Github Pages](https://jshwi.github.io/object_colors/build/html/index.html)  
[![Build Status](https://travis-ci.org/jshwi/object_colors.svg?branch=master)](https://travis-ci.org/jshwi/object_colors)
[![codecov.io](https://codecov.io/github/jshwi/object_colors/coverage.svg?branch=master)](https://codecov.io/github/jshwi/object_colors?branch=master)
[![PyPI version](https://badge.fury.io/py/object-colors.svg)](https://badge.fury.io/py/object-colors)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

    Instantiate with string for class colors
    
![alt text](./docs/_static/Screenshot%20from%202019-10-02%2012-22-07.png)

    Or set colors with integers

![alt text](./docs/_static/Screenshot%20from%202019-10-02%2012-23-05.png)

    The format doesn't matter
    
![alt text](./docs/_static/Screenshot%20from%202019-10-02%2012-24-03.png)

    Create subclasses

![alt text](./docs/_static/Screenshot%20from%202019-10-02%2012-28-55.png)

    Instead of using print() use get() to manipulate strings
    
![alt text](./docs/_static/Screenshot%20from%202019-10-02%2012-31-32.png)

    Subclass contained in self.__dict__

![alt text](./docs/_static/Screenshot%20from%202019-10-02%2014-02-53.png)


CHANGELOG.md
Changelog
=========
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

[Unreleased](https://github.com/jshwi/object-colors/compare/v2.3.1...HEAD)
------------------------------------------------------------------------

[2.3.1](https://github.com/jshwi/object-colors/releases/tag/v2.3.1) - 2023-09-24
------------------------------------------------------------------------
### Fixed
- add default `NoneType` to `__setattr__`

[2.3.0](https://github.com/jshwi/object-colors/releases/tag/v2.3.0) - 2023-09-24
------------------------------------------------------------------------
### Added
- Add `__all__`

### Changed
- update param passed to `Color.print`
- update param passed to `Color.get`

[2.2.0](https://github.com/jshwi/object-colors/releases/tag/v2.2.0) - 2023-01-04
------------------------------------------------------------------------
### Added
- Add py.typed

[2.1.0](https://github.com/jshwi/object-colors/releases/tag/v2.1.0) - 2022-05-11
------------------------------------------------------------------------
### Added
- Allows any object for `print`

[2.0.1](https://github.com/jshwi/object-colors/releases/tag/v2.0.1) - 2021-07-27
------------------------------------------------------------------------
### Security
Bump urllib3 from 1.26.3 to 1.26.5

[2.0.0](https://github.com/jshwi/object-colors/releases/tag/v2.0.0) - 2021-03-15
------------------------------------------------------------------------
### Added
- Adds ``colorama`` for ``Windows`` systems
- Overrides ``__repr__``
- Adds string formatter to ``get``
- Adds ``populate`` for all elements
- Raises ``TypeError`` if setting invalid attribute name
- Raises ``IndexError`` if setting of range ``int`` attributes
- Raises ``TypeError`` if setting invalid attribute type
- Raises ``ValueError`` if setting invalid attribute ``str`` value
- Overrides ``__len__``
- Allows empty ``get`` or ``print`` without raising an ``IndexError``

### Changed
- Renames: `` text`` -> ``fore``,  ``background`` -> ``back``
- Renames: `` purple`` -> ``magenta``
- Moves: ``_get_processed`` -> ``__setattr__``
- Improves compiled ANSI string and default types

### Deprecated
- ``populate_colors`` is deprecated in favour of ``populate("fore")``

### Removed
- Removes multicolor feature
- Removes ``pop`` method
- Removes functionality to pass single {1,2,3} digit number as {1,2,3} arg(s)

### Fixed
- Fixes ``print`` method to mirror builtin
- Removes default black background for none
- Adds all effects
- Objects can be properly added with ``__setattr__``

[v1.0.8](https://github.com/jshwi/object-colors/releases/tag/v1.0.8)  - 2020-02-02
------------------------------------------------------------------------
### Added
- Adds method to get multicoloured strings
- Makes some methods & variables public for more manual configuration

### Changed
- Updates README.rst

[v1.0.7](https://github.com/jshwi/object-colors/releases/tag/v1.0.7) - 2020-01-31
------------------------------------------------------------------------
### Added
- Adds ignore-case / scatter search
- Adds Individual word coloring
- All characters (including tildes / backslashes) searchable

### Removed
- Removes get_nested() method (now useless)

### Fixed
- "Ignore_case" fully functional

[v1.0.6](https://github.com/jshwi/object-colors/releases/tag/v1.0.6) - 2019-12-07
------------------------------------------------------------------------
### Changed
- Change ``populate_colors`` arg: "colors" -> populate=True

[v1.0.5](https://github.com/jshwi/object-colors/releases/tag/v1.0.5) - 2019-11-10
------------------------------------------------------------------------
#### Added
- Adds ``populate_colors`` method

[v1.0.4](https://github.com/jshwi/object-colors/releases/tag/v1.0.4) - 2019-11-09
------------------------------------------------------------------------
### Added
- Adds additional bold switch for more flexible use
- Adds unpack tuple feature to self.get()
- Adds full pop() return test module

### Changed
- Updates docs structure

[v1.0.3](https://github.com/jshwi/object-colors/releases/tag/v1.0.3) - 2019-10-08
------------------------------------------------------------------------
### Fix
- New Python unicode characters no longer cause build to fail

[v1.0.2](https://github.com/jshwi/object-colors/releases/tag/v1.0.2)  - 2019-10-07
------------------------------------------------------------------------
### Added
- Can now instantiate several subclasses at once
- Uses set() without overwriting previous values
- Made keys list global (class scope)
- Adds subclass tests

### Changed
- Updates docs

### Fixed
- Valid integers no longer missed in main loop

[v1.0.1](https://github.com/jshwi/object-colors/releases/tag/v1.0.1) - 2019-10-03
------------------------------------------------------------------------
### Added
- Adds delete keypair method
- Adds docs
- Adds setup.py
- Can now create separate color objects and add new ones
- Adds test suite

### Changed
- Updates README.md

### Fixed
- Resolves unresolved references/attributes


conftest.py
#!/usr/bin/env python3
from typing import List

from pytest import fixture

from object_colors import Color


@fixture
def color() -> Color:
    populated = Color()
    populated.populate_colors()
    return populated


@fixture
def color_str(color: Color, str_: str) -> Color:
    return color.green.get(str_)


@fixture
def str_() -> str:
    return "Cc: My Business <me@mybusiness.com>;"


@fixture
def marked(red: str, reset: str) -> str:
    return f"{red}Cc:{reset} My Business <me@mybusiness.com>;"


@fixture
def marked_color(green: str, red: str, reset: str) -> str:
    return f"{green}{red}Cc:{green} My Business <me@mybusiness.com>;{reset}"


@fixture
def marked_second(red: str, reset: str) -> str:
    return f"Cc: {red}My{reset} Business <me@mybusiness.com>;"


@fixture
def marked_second_color(green: str, red: str, reset: str) -> str:
    return f"{green}Cc: {red}My{green} Business <me@mybusiness.com>;{reset}"


@fixture
def all_cs(red: str, reset: str) -> str:
    return (
        f"{red}C{reset}{red}c{reset}{red}:{reset} My "
        f"Business <me@mybusiness.{red}c{reset}om>;"
    )


@fixture
def dupe_marked_color(green: str, red: str, reset: str) -> str:
    """do dupe fixture but with ignore"""
    return (
        f"{green}This is a string that says {red}one{green} several times. "
        f"It says {red}one{green} in this sentence. And {red}one{green} in "
        f"this sentence. This sentence also has {red}one{green} in it. Lastly "
        f"this sentence will also say {red}one{green}{reset}"
    )


@fixture
def dupe_marked(green: str, red: str, reset: str) -> str:
    """do dupe fixture but with ignore"""
    return (
        f"This is a string that says {red}one{reset} several times. It says "
        f"{red}one{reset} in this sentence. And {red}one{reset} in this "
        f"sentence. This sentence also has {red}one{reset} in it. Lastly this "
        f"sentence will also say {red}one{reset}"
    )


@fixture
def all_cs_color(red: str, green: str, reset: str) -> str:
    return (
        f"{green}{red}C{green}{red}c{green}{red}:{green} My "
        f"Business <me@mybusiness.{red}c{green}om>;{reset}"
    )


@fixture
def all_cs_no_caps(red: str, reset: str) -> str:
    return (
        f"C{red}c{reset}{red}:{reset} My "
        f"Business <me@mybusiness.{red}c{reset}om>;"
    )


@fixture
def all_cs_no_caps_color(red: str, green: str, reset: str) -> str:
    return (
        f"{green}C{red}c{green}{red}:{green} My "
        f"Business <me@mybusiness.{red}c{green}om>;{reset}"
    )


@fixture
def all_cs_no_caps_blue(blue: str, reset: str) -> str:
    return (
        f"C{blue}c{reset}{blue}:{reset} My "
        f"Business <me@mybusiness.{blue}c{reset}om>;"
    )


@fixture
def all_cs_no_caps_blue_color(blue: str, green: str, reset: str) -> str:
    return (
        f"{green}C{blue}c{green}{blue}:{green} My "
        f"Business <me@mybusiness.{blue}c{green}om>;{reset}"
    )


@fixture
def exact_idx(red: str, reset: str) -> str:
    return (
        f"{red}C{reset}{red}c{reset}: My Business "
        f"<me@mybusiness.{red}c{reset}om>;"
    )


@fixture
def exact_idx_color(green: str, red: str, reset: str) -> str:
    return (
        f"{green}{red}C{green}{red}c{green}: My Business "
        f"<me@mybusiness.{red}c{green}om>;{reset}"
    )


@fixture
def dupes() -> str:
    return (
        "This is a string that says one several times. It says one in this "
        "sentence. And one in this sentence. This sentence also has one in "
        "it. Lastly this sentence will also say one"
    )


@fixture
def colored_dupes() -> str:
    return (
        "\u001b[0;32;40mThis is a string that says one several times. It says "
        "one in this sentence. And one in this sentence. This sentence also "
        "has one in it. Lastly this sentence will also say one\u001b[0;0m"
    )


@fixture
def scatter_cs(green: str, red: str, reset: str) -> str:
    return (
        f"{red}C{reset}{red}c{reset}{red}:{reset} My Business "
        f"<me@mybusiness.{red}c{reset}om>;"
    )


@fixture
def scatter_cs_color(green: str, red: str, reset: str) -> str:
    return (
        f"{green}{red}C{green}{red}c{green}{red}:{green} My Business "
        f"<me@mybusiness.{red}c{green}om>;{reset}"
    )


@fixture
def scatter_cs_blue(green: str, blue: str, reset: str) -> str:
    return (
        f"{blue}C{reset}{blue}c{reset}{blue}:{reset} My Business "
        f"<me@mybusiness.{blue}c{reset}om>;"
    )


@fixture
def scatter_cs_blue_color(green: str, blue: str, reset: str) -> str:
    return (
        f"{green}{blue}C{green}{blue}c{green}{blue}:{green} My Business "
        f"<me@mybusiness.{blue}c{green}om>;{reset}"
    )


@fixture
def scatter_cs_exact(green: str, red: str, reset: str) -> str:
    return f"C{red}c{reset}: My Business <me@mybusiness.{red}c" f"{reset}om>;"


@fixture
def scatter_cs_exact_color(green: str, red: str, reset: str) -> str:
    return (
        f"{green}C{red}c{green}: My Business "
        f"<me@mybusiness.{red}c{green}om>;{reset}"
    )


@fixture
def spaced_words(green: str, red: str, reset: str) -> str:
    return f"{red}Cc:{reset} My {red}Business{reset} <me@mybusiness.com>;"


@fixture
def spaced_words_color(green: str, red: str, reset: str) -> str:
    return (
        f"{green}{red}Cc:{green} My {red}Business{green} "
        f"<me@mybusiness.com>;{reset}"
    )


@fixture
def green() -> str:
    return "\u001b[0;32;40m"


@fixture
def red() -> str:
    return "\u001b[0;31;40m"


@fixture
def blue() -> str:
    return "\u001b[0;34;40m"


@fixture
def reset() -> str:
    return "\u001b[0;0m"


@fixture
def color_keys(color: Color, str_: str) -> str:
    return color.red.get_key(str_, "c", ignore_case=True, scatter=True)


@fixture
def attrs() -> List[str]:
    return [
        "get",
        "set",
        "get",
        "get_key",
        "print",
        "print_key",
        "pop",
        "multicolor",
        "populate_colors",
    ]


@fixture
def colors():
    return [
        "black",
        "red",
        "green",
        "yellow",
        "blue",
        "purple",
        "cyan",
        "white",
    ]


@fixture
def instances(colors) -> List[str]:
    return ["text", "effect", "background", "bold"] + colors


@fixture
def for_multiple_colors():
    return "Testing for multiple colors"


@fixture
def codes():
    return [
        "0;31;40m",
        "0;32;40m",
        "0;33;40m",
        "0;34;40m",
        "0;35;40m",
        "0;36;40m",
        "0;37;40m",
    ]


@fixture
def long_str():
    return (
        "Beautiful is better than ugly."
        "Explicit is better than implicit."
        "Simple is better than complex."
        "Complex is better than complicated."
        "Flat is better than nested."
        "Sparse is better than dense."
        "Readability counts."
        "Special cases aren't special enough to break the rules."
        "Although practicality beats purity."
        "Errors should never pass silently."
        "Unless explicitly silenced."
        "In the face of ambiguity, refuse the temptation to guess."
        "There should be one-- and preferably only one --obvious way to do it."
        "Although that way may not be obvious at first unless you're Dutch."
        "Now is better than never."
        "Although never is often better than *right* now."
        "If the implementation is hard to explain, it's a bad idea."
        "If the implementation is easy to explain, it may be a good idea."
        "Namespaces are one honking great idea -- let's do more of those!"
    )
